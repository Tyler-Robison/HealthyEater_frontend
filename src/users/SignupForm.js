import React, { useContext, useEffect } from "react";
import UserAPI from "../APIs/userAPI";
import { useNavigate } from 'react-router-dom';
import { useFormik } from "formik";
import signupValidate from "./user_validators/signupValidate";
import GlobalContext from "../context/GlobalContext";
import useTimedMessage from "../customHooks/useTimedMessage";
import './SignupForm.css'

/** SignupForm allows users to create new accounts
 * 
 * upon succesful registration currentUser will be created, allowing all components to access user info*/
const SignupForm = () => {
    const validate = signupValidate
    const navigate = useNavigate();
    const { currentUser, login } = useContext(GlobalContext)
    const [failureMsg, setfailureMsg] = useTimedMessage()

    useEffect(() => {
        if (currentUser) navigate('/')
    }, [])

    const formik = useFormik({
        initialValues: {
            username: '',
            password: ''
        },
        validate,
        onSubmit: values => signup(values),
    })

    // login sets token, which triggers useEffect in App.js
    const signup = async (values) => {
        try {
            const res = await UserAPI.register(values)
            login(res.token);
            formik.resetForm();
            navigate('/');
        } catch (err) {
            setfailureMsg(true)
            formik.resetForm();
        }
    }

    return (
        <div className="SignupForm">
            <div className="row mt-3">
                <div className="SignupForm-div col-12">

                    <h1>Enter Username/Password</h1>
                    {failureMsg && <p>That username is already taken</p>}
                    <form onSubmit={formik.handleSubmit}>
                        <div>
                            <div><label htmlFor="username">Username</label></div>
                            <input
                                id="username"
                                name="username"
                                type='text'
                                onChange={formik.handleChange}
                                onBlur={formik.handleBlur}
                                value={formik.values.username}
                                placeholder="Username"
                            />
                            {formik.touched.username && formik.errors.username && (
                                <div>{formik.errors.username}</div>
                            )}
                        </div>
                        <div>
                            <div><label htmlFor="password">Password</label></div>
                            <input
                                id="password"
                                name="password"
                                type='password'
                                onChange={formik.handleChange}
                                onBlur={formik.handleBlur}
                                value={formik.values.password}
                                placeholder="Password"
                            />
                            {formik.touched.password && formik.errors.password && (
                                <div>{formik.errors.password}</div>
                            )}
                        </div>
                        <button className="general-btn SignupForm-btn" type="submit">Sign-up</button>
                    </form>
                </div>
            </div>
        </div>
    )
}

export default SignupForm;