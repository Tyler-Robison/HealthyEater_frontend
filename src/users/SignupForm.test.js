import { render } from '@testing-library/react';
import SignupForm from './SignupForm';
import { MemoryRouter } from "react-router-dom";
import ContextProvider from '../testContext'


it("renders without crashing", function () {
    <MemoryRouter>
        <ContextProvider>
            <SignupForm />
        </ContextProvider>
    </MemoryRouter>
});

it("matches snapshot", function () {

    const container = render(
        <MemoryRouter>
            <ContextProvider>
                <SignupForm />
            </ContextProvider>
        </MemoryRouter>
    );
    expect(container.asFragment()).toMatchSnapshot();
});