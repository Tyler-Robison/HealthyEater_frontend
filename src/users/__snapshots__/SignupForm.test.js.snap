// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="SignupForm"
  >
    <div
      class="SignupForm-div"
    >
      <h1>
        Enter Username/Password
      </h1>
      <form>
        <div>
          <div>
            <label
              for="username"
            >
              Username
            </label>
          </div>
          <input
            id="username"
            name="username"
            placeholder="Username"
            type="text"
            value=""
          />
        </div>
        <div>
          <div>
            <label
              for="password"
            >
              Password
            </label>
          </div>
          <input
            id="password"
            name="password"
            placeholder="Password"
            type="password"
            value=""
          />
        </div>
        <button
          class="general-btn SignupForm-btn"
          type="submit"
        >
          Sign-up
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
